### PR TITLE
Sparse CSR CUDA: add `addmv_out`

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #61858
* #61857
* #65606
* #64391
* #63948
* #63511
* #63510
* #63509
* #61536
* #66485
* __->__ #61407

This PR adds `addmv_out_sparse_csr_cuda`. The operation is used to
compute matrix-vector multiplication. Since structured_delegate is used
we only need to implement the out variant, the in-place and normal
variants are autogenerated.
Working on this PR revealed that float16 (and probably bfloat16) inputs
do not work correctly in cusparse, therefore for this case `addmm` is
used with squeezes and unsqueezes.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel

Differential Revision: [D31584499](https://our.internmc.facebook.com/intern/diff/D31584499)

### CODE REVIEW

**Review by xwang233 — COMMENTED** [review]
Thanks for the work on this! Is FP16 and BF16 of `cusparseSpMV` still an issue? If so, a C++ repro would be much helpful. Everything else LGTM.

File: aten/src/ATen/native/sparse/cuda/SparseBlas.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          const_cast<Tensor&>(result),
+          self,
+          at::native::scalar_tensor(
+              beta, self.scalar_type(), c10::nullopt /* layout */, at::kCPU, c10::nullopt /* pin_memory */));
+    }
+  }
+
+  // cuda 11.3 version computes garbage for float16 inputs
+  // couldn't check bfloat16 because it requires Ampere GPU but I assume the problem is same
+  // addmm works fine
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply aten/src/ATen/native/sparse/cuda/SparseBlas.cpp:48]: I'll prepare a standalone C file and test it in CUDA 11.4.

File: aten/src/ATen/native/Blas.cpp [hunk 2]
```diff
@@ -16,7 +16,7 @@ TORCH_META_FUNC(addmv)(const Tensor &self, const Tensor &mat, const Tensor &vec,
   TORCH_CHECK(mat.size(1) == vec.size(0) && (mat.size(0) == self.numel() || self.numel() == 1),
      "size mismatch, got ", self.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
   auto names = at::namedinference::propagate_names_for_addmv(mat, vec, self);
-  set_output(0, IntArrayRef(mat.sizes().data(), 1), {}, mat.options(), names);
+  set_output(0, IntArrayRef(mat.sizes().data(), 1), {}, vec.options(), names);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/Blas.cpp:19]: Why this change?

Comment (IvanYashchuk) [reply aten/src/ATen/native/Blas.cpp:19]: Because `mat` can be now sparse csr tensor while we need strided tensor as output. `vec` is a strided tensor so its options are used to create the output tensor.

File: aten/src/ATen/cuda/CUDASparseDescriptors.h [hunk 3]
```diff
@@ -42,6 +42,12 @@ class TORCH_CUDA_CPP_API CuSparseDnMatDescriptor
   CuSparseDnMatDescriptor(const Tensor& input);
 };
 
+class TORCH_CUDA_CPP_API CuSparseDnVecDescriptor
+    : public CuSparseDescriptor<cusparseDnVecDescr, &cusparseDestroyDnVec> {
+ public:
+  CuSparseDnVecDescriptor(const Tensor& input);
```
Comment (cpuhrsch) [inline aten/src/ATen/cuda/CUDASparseDescriptors.h:48]: In general you might want to make this constructors explicit. To quote our linter "A single-argument constructor may inadvertently be used as a type conversion constructor."

**Comment by IvanYashchuk** [comment]
@ngimel, could you please review and help merge this stack starting from this PR?

**Review by ngimel — COMMENTED** [review]
This looks good, I've left minor comments. Do you know if there's an advantage to using cusparseSpMV compared to cusparseSpMM (that is used anyway for bfloat16/half)?

**Comment by ngimel** [comment]
There are also real unused variable errors on this PR (probably not caused by it, and just require a rebase), unfortunately they are too hard to find in the logs 
```
2021-10-11T06:27:54.9960767Z [01m[K/var/lib/jenkins/workspace/torch/csrc/generic/StorageMethods.cpp:[m[K In function '[01m[KPyObject* THPQUInt2x4Storage_new(PyObject*, PyObject*)[m[K':
2021-10-11T06:27:54.9963406Z [01m[K/var/lib/jenkins/workspace/torch/csrc/generic/StorageMethods.cpp:62:8:[m[K [01;31m[Kerror: [m[Kunused variable '[01m[Kself[m[K' [[01;31m[K-Werror=unused-variable[m[K]
2021-10-11T06:27:54.9965138Z    auto [01;31m[Kself[m[K = (THPStorage*)_self;
2021-10-11T06:27:54.9966001Z         [01;31m[K^~~~[m[K
2021-10-11T06:27:54.9967879Z [01m[K/var/lib/jenkins/workspace/torch/csrc/generic/StorageMethods.cpp:[m[K In function '[01m[KPyObject* THPQUInt2x4Storage_newWithFile(PyObject*, PyObject*)[m[K':
2021-10-11T06:27:54.9970722Z [01m[K/var/lib/jenkins/workspace/torch/csrc/generic/StorageMethods.cpp:289:13:[m[K [01;31m[Kerror: [m[Kunused variable '[01m[Kfd_obj[m[K' [[01;31m[K-Werror=unused-variable[m[K]
2021-10-11T06:27:54.9972337Z    PyObject *[01;31m[Kfd_obj[m[K = PyTuple_GetItem(args, 0);
2021-10-11T06:27:54.9973178Z              [01;31m[K^~~~~~[m[K
2021-10-11T06:27:54.9974063Z cc1plus: all warnings being treated as errors
```

**Comment by IvanYashchuk** [comment]
> @IvanYashchuk what are the bugs with bfloat16/float16 mv? cc @xwang233 to follow up with cusparse

Let's take the code sample for SpMV CSR from https://github.com/NVIDIA/CUDALibrarySamples/tree/master/cuSPARSE/spmv_csr
Modifying slightly the code to use `__half` dtype computes:
```
spmv_csr_example test FAILED: wrong result
33664.000000 2.250000 -0.000157 18.000000
```
or with `__nv_bfloat16`:
```
spmv_csr_example test FAILED: wrong result
212676479325586539664609129644855132160.000000 4.000000 -0.000000 32.000000
```
while the expected result is `19.0f, 8.0f, 51.0f, 52.0f`. This is tested with CUDA 11.4.2

The code to reproduce the problem is [here](https://gist.github.com/IvanYashchuk/a33ae7d01ff1a09e97523b13f55b9216).

**Review by ngimel — APPROVED** [review]
